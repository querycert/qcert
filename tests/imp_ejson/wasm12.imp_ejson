/* cast objects */
module main {
  wasm12(x) return y {
    let a0 := { $class: ["Address"], $data: { city : "xy" }};
    let a1 := { $class: ["USAddress"], $data: { city : "NY", zip : 10001 }};
    let p0 := { $class: ["Person"], $data: { name: "donald" }} ;
    let y :=
    /* valid casts */
      [[ Runtime.cast(["Address"], a0)
       , Runtime.cast(["USAddress"], a1)
       , Runtime.cast(["Address"], a1)
       , Runtime.cast(["Object"], a0)
       , Runtime.cast(["Object"], a1)
       , Runtime.cast(["Object"], p0)
       ]
    /* invalid casts */
      ,[ Runtime.cast(["Address"], p0)
       , Runtime.cast(["Person"], a0)
       , Runtime.cast(["Person"], a1)
       , Runtime.cast(["USAddress"], a0)
      ]];
  }
}
