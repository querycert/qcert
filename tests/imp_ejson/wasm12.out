[
  [
    {
      "$left": {
        "$class": [ "Address" ],
        "$data": { "city": "xy" }
      }
    },
    {
      "$left": {
        "$class": [ "USAddress" ],
        "$data": { "city": "NY", "zip": { "$nat": 10001 } }
      }
    },
    {
      "$left": {
        "$class": [ "Address" ],
        "$data": { "city": "NY", "zip": { "$nat": 10001 } }
      }
    },
    {
      "$left": { "$class": [ "Object" ], "$data": { "city": "xy" } }
    },
    {
      "$left": {
        "$class": [ "Object" ],
        "$data": { "city": "NY", "zip": { "$nat": 10001 } }
      }
    },
    { "$left": { "$class": [ "Object" ], "$data": { "name": "donald" } } }
  ],
  [ { "$right": null },
    { "$right": null },
    { "$right": null },
    { "$right": null } ]
]
